### DISCUSSION TITLE
How do we build and test Release versions for iOS in Maui now that Visual Studio for Mac is dead?

### DISCUSSION THREAD

**Discussion by jonmdev** [dotnet/maui#23492] [Q&A] [answered]
To build to iOS in Debug mode, one can attach an iPhone by USB to a Windows computer and in Debug mode select the iPhone. This works easily enough.

But how are we supposed to build to iOS in Release mode? I see Visual Studio has been discontinued for Macs: https://devblogs.microsoft.com/visualstudio/visual-studio-for-mac-retirement-announcement/

## Ideas
I found this page that suggests two methods:

1) Install Dot Net on the Mac and run the build from a command prompt: 
https://learn.microsoft.com/en-us/dotnet/maui/ios/cli?view=net-maui-8.0

2) Use "Pair To Mac" and run the build from the Windows computer:
https://learn.microsoft.com/en-us/dotnet/maui/ios/pair-to-mac?view=net-maui-8.0

Are both of these still working? I presume "Pair to Mac" is how most people are doing it then? So you don't have to clone the project over to the Mac?

## Pair to Mac?
The page on "Pair To Mac" says:

>Before following the instructions in this article, on a Mac, [install Xcode](https://apps.apple.com/us/app/xcode/id497799835?mt=12). Then manually open Xcode, after installation, so that it can add additional components. In addition, you should also install either the latest [Visual Studio 2022 for Mac](https://visualstudio.microsoft.com/vs/mac/) or [Mono](https://www.mono-project.com/download/stable/#download-mac). In addition, if you have a Mac computer with Apple silicon please ensure that Rosetta is installed.

So since we don't have Visual Studio 2022 for Mac anymore, does this work by installing Mono & XCode still? Is this the correct working method still?

## Copy to iPhone and Run?
If you do succeed in the above, how do you then copy the build to an iPhone and run it, since iOS doesn't allow "sideloading"?

Thanks for any help.

**Comment by michael-senva** [comment, marked as answer]
Pair to Mac with Mono/Xcode is the workflow I've been using for several years now for debug and Release modes. I can also confirm that it is working in .NET 8 after just recently migrating from .NET 7. The process facilitates automatic deployment to the phone and should run after finishing with the app build still being present on the device until modified/removed manually.

FWIW, I only push to the App Store via the organizer in Xcode, I haven't tried the manual packaging process.

EDIT: I didn't realize you could push debug builds to the iOS device connected via USB to the windows PC. I've always had it connected to the Mac, may be why I've never seen an issue between the two modes.

**Comment by ToolmakerSteve** [comment]
It is possible now to develop iOS directly on Mac using `Visual Studio Code` with C# and .Net extensions/downloads. Open [Getting Started with C# in VS Code](https://code.visualstudio.com/docs/csharp/get-started) on your Mac. The various links should offer downloads for macOS (e.g., Microsoft Marketplace detects your platform).

CAVEAT: I haven't tested MAUI this way; I've been doing that by connecting from VS 2022 on Windows.

(I have my iPad connected to my mac.)

**Comment by JoacimWall** [comment]
One easy solution is to use Azure pipeline and build your app.  Then you will have good control of all different version for MAUI, Dotnet, Workload, Xcode and more.

Below is sample of maui pipeline

resources:

trigger: none
stages:
- stage: __default
  variables: 
  - group: Mobile.App.Prod
  jobs:
  - job: Job
    pool:
      vmImage: 'macOS-15'
    steps:
      - checkout: self
      - task: UseDotNet@2
        displayName: 'Use .Net 9'
        inputs:
         version: 9.0.202
         includePreviewVersions: false
    
      - task: CmdLine@2
        displayName: 'Install Maui workload'
        inputs:
          script: 'dotnet workload install maui --version 9.0.202'

      - task: InstallAppleCertificate@2
        displayName: 'Install Apple cert'
        inputs:
          certSecureFile: 'xxxxxx_Certificate_Nov_14_25.p12'
          certPwd: '$(IOS_Cert_Password)'
      - task: InstallAppleProvisioningProfile@1
        displayName: 'Install an Apple provisioning profile'
        name: ProvisioningProfileTask_App_Store
        inputs:
         provProfileSecureFile: 'xxxxxxxxxx_Appstore.mobileprovision'
      - task: DownloadSecureFile@1
        name: keystore
        displayName: Download Android Keystore 
        inputs:
          secureFile: 'xxxxxxx-upload-keystore.jks'
      - script: |
         echo Downloaded xxxxxxx-upload-keystore.jks
         echo Working Directory $(System.DefaultWorkingDirectory)\\xxxxxx-upload-keystore.jks
         mv $(keystore.secureFilePath) $(System.DefaultWorkingDirectory)
        displayName: Move Keystore to Working Directory
      - task: Bash@3
        displayName: Build Ios App
        inputs:
          targetType: 'inline'
          script: |
            sudo xcode-select -switch /Applications/Xcode_16.2.app
            cd 'xxxxxxx App/scr/xxxxxx'
            dotnet publish xxxxxx.csproj -f net9.0-ios -c Release -p:ApplicationId=com.xxxxx.xxxx.xxx.xxxx -p:RuntimeIdentifier=ios-arm64 -p:BuildIpa=True -o ./artifacts
      - task: CopyFiles@2
        inputs:
          SourceFolder: '$(Agent.BuildDirectory)'
          Contents: '**/*.ipa'
          TargetFolder: '$(Build.ArtifactStagingDirectory)'
          flattenFolders: true
      - task: CopyFiles@2
        inputs:
          SourceFolder: '$(Agent.BuildDirectory)'
          Contents: '**/*.dSYM/**'
          TargetFolder: '$(Build.ArtifactStagingDirectory)'
          flattenFolders: false
      - task: Bash@3
        displayName: Build Android App
        inputs:
          targetType: 'inline'
          script: |
            cd 'xxxxxx App/scr/xxxxxxx'
            dotnet publish xxxxxx.csproj -f net9.0-android -c Release -p:ApplicationId=com.xxxxx.xxxxx.xxxxx.xxxxx -p:RuntimeIdentifier=android-arm64 -p:AndroidSigningKeyPass=$(Android_Key_Password) -p:AndroidSigningStorePass=$(Android_Keystore_Password) /p:AndroidSigningKeyStore=$(System.DefaultWorkingDirectory)\\xxxxx-upload-keystore.jks -p:AndroidSigningKeyAlias=$(Android_Key_Alias) -p:AndroidKeyStore=true -p:AndroidEnableProfiler=true
      - task: CopyFiles@2
        inputs:
          SourceFolder: '$(Agent.BuildDirectory)'
          Contents: '**/*.aab'
          TargetFolder: '$(Build.ArtifactStagingDirectory)'
          flattenFolders: true
      - task: PublishBuildArtifacts@1
        inputs:
          PathtoPublish: '$(Build.ArtifactStagingDirectory)'
          ArtifactName: 'drop'
          publishLocation: 'Container'
